From a27ff7c9a6f334a7bd7c1e0b5f772df1ba56f7d0 Mon Sep 17 00:00:00 2001
From: Gu Zheng <guzheng1@huawei.com>
Date: Fri, 5 Nov 2021 12:58:47 -0500
Subject: [PATCH 2/2] tmpfs: clear S_ISGID when setting posix ACLs

commit 497de07d89c1410d76a15bec2bb41f24a2a89f31 upstream.

This change was missed the tmpfs modification in In CVE-2016-7097
commit 073931017b49 ("posix_acl: Clear SGID bit when setting
file permissions")
It can test by xfstest generic/375, which failed to clear
setgid bit in the following test case on tmpfs:

  touch $testfile
  chown 100:100 $testfile
  chmod 2755 $testfile
  _runas -u 100 -g 101 -- setfacl -m u::rwx,g::rwx,o::rwx $testfile

Committer's note: This was added from the OnePlus 3's LineageOS kernel
tree, commit 1026a8ce. As far as I can tell, this is
Co-Authored-By: Willy Tarreau <w@1wt.eu>

Signed-off-by: Gu Zheng <guzheng1@huawei.com>
Signed-off-by: Al Viro <viro@zeniv.linux.org.uk>
Signed-off-by: Jan Kara <jack@suse.cz>
Signed-off-by: Willy Tarreau <w@1wt.eu>
Change-Id: I140a8f4b3a29fbab12f163709c3c1c21241a9e98
---
 fs/generic_acl.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/fs/generic_acl.c b/fs/generic_acl.c
index e9c0746e8205..a614e8581a33 100644
--- a/fs/generic_acl.c
+++ b/fs/generic_acl.c
@@ -82,14 +82,20 @@ generic_acl_set(struct dentry *dentry, const char *name, const void *value,
 			return PTR_ERR(acl);
 	}
 	if (acl) {
+		struct posix_acl *old_acl;
+
 		error = posix_acl_valid(acl);
 		if (error)
 			goto failed;
 		switch (type) {
 		case ACL_TYPE_ACCESS:
-			error = posix_acl_update_mode(inode, &inode->i_mode, &acl);
+			old_acl = acl;
+			error = posix_acl_update_mode(inode, &inode->i_mode,
+						      &acl);
 			if (error)
 				goto failed;
+			if (!acl)
+				posix_acl_release(old_acl);
 			inode->i_ctime = CURRENT_TIME;
 			break;
 		case ACL_TYPE_DEFAULT:
-- 
2.25.1

